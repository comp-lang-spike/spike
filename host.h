
#ifndef __spk_host_h__
#define __spk_host_h__


#include "om.h"
#include <stdarg.h>
#include <stdio.h>


/****/ void SpkHost_Init(void);


/* halting */
/****/ void SpkHost_Halt(int, const char *);
/****/ void SpkHost_HaltWithString(int, SpkUnknown *);
/****/ void SpkHost_VHaltWithFormat(int, const char *, va_list);


/* base types */
/*****/ int SpkHost_IsInteger(SpkUnknown *);
/*****/ int SpkHost_IsString(SpkUnknown *);
/****/ long SpkHost_IntegerAsCLong(SpkUnknown *);
SpkUnknown *SpkHost_IntegerFromCLong(long);
SpkUnknown *SpkHost_FloatFromCDouble(double);
/***/ char *SpkHost_StringAsCString(SpkUnknown *);
SpkUnknown *SpkHost_StringFromCString(const char *);
SpkUnknown *SpkHost_StringFromCStringAndLength(const char *, size_t);
SpkUnknown *SpkHost_StringConcat(SpkUnknown **, SpkUnknown *);


/* symbols */
/*****/ int SpkHost_IsSymbol(SpkUnknown *);
/***/ char *SpkHost_SymbolAsCString(SpkUnknown *);
SpkUnknown *SpkHost_SymbolFromString(const char *);
SpkUnknown *SpkHost_SymbolFromCStringAndLength(const char *, size_t);


/* selectors */
/***/ char *SpkHost_SelectorAsCString(SpkUnknown *);
SpkUnknown *SpkHost_NewKeywordSelectorBuilder(void);
/****/ void SpkHost_AppendKeyword(SpkUnknown **, SpkUnknown *);
SpkUnknown *SpkHost_GetKeywordSelector(SpkUnknown *, SpkUnknown *);


/* symbol dictionaries */
SpkUnknown *SpkHost_NewSymbolDict(void);
/****/ void SpkHost_DefineSymbol(SpkUnknown *, SpkUnknown *, SpkUnknown *);
SpkUnknown *SpkHost_SymbolValue(SpkUnknown *, SpkUnknown *);
SpkUnknown *SpkHost_FindSymbol(SpkUnknown *, SpkUnknown *);
/*****/ int SpkHost_NextSymbol(SpkUnknown *, size_t *,
                               SpkUnknown **, SpkUnknown **);


/* literal dictionaries */
SpkUnknown *SpkHost_NewLiteralDict(void);
unsigned int SpkHost_InsertLiteral(SpkUnknown *, SpkUnknown *);


/* arguments */
/*****/ int SpkHost_IsArgs(SpkUnknown *);
/**/ size_t SpkHost_ArgsSize(SpkUnknown *);
SpkUnknown *SpkHost_GetArg(SpkUnknown *, size_t);
SpkUnknown *SpkHost_EmptyArgs(void);
SpkUnknown *SpkHost_ArgsFromVAList(va_list);
SpkUnknown *SpkHost_GetArgs(SpkUnknown **, size_t,
                            SpkUnknown *, size_t);

/* i/o */
/*****/ int SpkHost_IsFileStream(SpkUnknown *);
/***/ FILE *SpkHost_FileStreamAsCFileStream(SpkUnknown *);


/* as yet unclassified */
SpkUnknown *SpkHost_ObjectAsString(SpkUnknown *);
/****/ void SpkHost_PrintObject(SpkUnknown *, FILE *);


#endif /* __spk_host_h__ */
