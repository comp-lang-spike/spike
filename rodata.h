
#ifndef __rodata_h__
#define __rodata_h__


#include "obj.h"
#include "oper.h"


/* symbols */
extern struct Symbol *self;


/* selectors */
extern struct Symbol *__add__;
extern struct Symbol *__addr__;
extern struct Symbol *__apply__;
extern struct Symbol *__apply__;
extern struct Symbol *__band__;
extern struct Symbol *__bneg__;
extern struct Symbol *__bor__;
extern struct Symbol *__bxor__;
extern struct Symbol *__div__;
extern struct Symbol *__eq__;
extern struct Symbol *__ge__;
extern struct Symbol *__gt__;
extern struct Symbol *__ind__;
extern struct Symbol *__index__;
extern struct Symbol *__le__;
extern struct Symbol *__lneg__;
extern struct Symbol *__lshift__;
extern struct Symbol *__lt__;
extern struct Symbol *__mod__;
extern struct Symbol *__mul__;
extern struct Symbol *__ne__;
extern struct Symbol *__neg__;
extern struct Symbol *__pos__;
extern struct Symbol *__pred__;
extern struct Symbol *__rshift__;
extern struct Symbol *__str__;
extern struct Symbol *__sub__;
extern struct Symbol *__succ__;
extern struct Symbol *_import;
extern struct Symbol *_xinit;
extern struct Symbol *_predef;
extern struct Symbol *_thunk;
extern struct Symbol *badExpr;
extern struct Symbol *blockCopy;
extern struct Symbol *cannotReenterBlock;
extern struct Symbol *cannotReturn;
extern struct Symbol *compoundExpression;
extern struct Symbol *doesNotUnderstand;
extern struct Symbol *failOnError;
extern struct Symbol *importModule;
extern struct Symbol *importPackage;
extern struct Symbol *xinit;
extern struct Symbol *xmain;
extern struct Symbol *mustBeBoolean;
extern struct Symbol *mustBeSymbol;
extern struct Symbol *mustBeTuple;
extern struct Symbol *new;
extern struct Symbol *noRunnableFiber;
extern struct Symbol *numArgs;
extern struct Symbol *printString;
extern struct Symbol *recursiveDoesNotUnderstand;
extern struct Symbol *redefinedSymbol;
extern struct Symbol *source;
extern struct Symbol *undefinedSymbol;
extern struct Symbol *unknownOpcode;
extern struct Symbol *wrongNumberOfArguments;

/* selectors -- class names */
extern struct Symbol *Boolean;
extern struct Symbol *False;
extern struct Symbol *True;

/* selectors -- parser */
extern struct Symbol *concat;
extern struct Symbol *declSpecs;
extern struct Symbol *exprAnd;
extern struct Symbol *exprAssign;
extern struct Symbol *exprAttr;
extern struct Symbol *exprAttrVar;
extern struct Symbol *exprBinaryOp;
extern struct Symbol *exprBlock;
extern struct Symbol *exprCall;
extern struct Symbol *exprCompound;
extern struct Symbol *exprCond;
extern struct Symbol *exprId;
extern struct Symbol *exprKeyword;
extern struct Symbol *exprLiteral;
extern struct Symbol *exprName;
extern struct Symbol *exprNI;
extern struct Symbol *exprOr;
extern struct Symbol *exprPostOp;
extern struct Symbol *exprPreOp;
extern struct Symbol *exprUnaryOp;
extern struct Symbol *isSpec;
extern struct Symbol *left;
extern struct Symbol *next;
extern struct Symbol *nextArg;
extern struct Symbol *operAdd;
extern struct Symbol *operAddr;
extern struct Symbol *operApply;
extern struct Symbol *operBAnd;
extern struct Symbol *operBNeg;
extern struct Symbol *operBOr;
extern struct Symbol *operBXOr;
extern struct Symbol *operDiv;
extern struct Symbol *operEq;
extern struct Symbol *operGE;
extern struct Symbol *operGT;
extern struct Symbol *operInd;
extern struct Symbol *operIndex;
extern struct Symbol *operLE;
extern struct Symbol *operLNeg;
extern struct Symbol *operLShift;
extern struct Symbol *operLT;
extern struct Symbol *operMod;
extern struct Symbol *operMul;
extern struct Symbol *operNE;
extern struct Symbol *operNeg;
extern struct Symbol *operPos;
extern struct Symbol *operPred;
extern struct Symbol *operRShift;
extern struct Symbol *operSub;
extern struct Symbol *operSucc;
extern struct Symbol *stmtBreak;
extern struct Symbol *stmtCompound;
extern struct Symbol *stmtContinue;
extern struct Symbol *stmtDefClass;
extern struct Symbol *stmtDefMethod;
extern struct Symbol *stmtDefModule;
extern struct Symbol *stmtDefSpec;
extern struct Symbol *stmtDefVar;
extern struct Symbol *stmtDoWhile;
extern struct Symbol *stmtExpr;
extern struct Symbol *stmtFor;
extern struct Symbol *stmtIfElse;
extern struct Symbol *stmtPragmaSource;
extern struct Symbol *stmtReturn;
extern struct Symbol *stmtWhile;
extern struct Symbol *stmtYield;
extern struct Symbol *symbolNodeForSymbol;

typedef struct SpecialSelector {
    struct Symbol **selector;
    size_t argumentCount;
} SpecialSelector;

extern SpecialSelector operSelectors[NUM_OPER];
extern SpecialSelector operCallSelectors[NUM_CALL_OPER];


/* integers */
extern struct Integer *zero;
extern struct Integer *one;


/* strings */
extern struct String *emptyString;
extern struct String *unknownSelector;
extern struct String *unknownSourcePathname;


/* arguments */
extern struct Array *emptyArgs;


int InitSymbols(void);
int InitReadOnlyData(void);
struct Symbol *ParseSelector(const char *);


#endif /* __rodata_h__ */
