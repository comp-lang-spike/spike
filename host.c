
#include "host.h"

#include "array.h"
#include "behavior.h"
#include "dict.h"
#include "float.h"
#include "int.h"
#include "interp.h"
#include "io.h"
#include "native.h"
#include "rodata.h"
#include "str.h"
#include "sym.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>


/****/ void SpkHost_Init(void) {
}


/*------------------------------------------------------------------------*/
/* halting */

static const char *haltDesc(int code) {
    const char *desc;
    
    switch (code) {
    case Spk_HALT_ASSERTION_ERROR: desc = "assertion error"; break;
    case Spk_HALT_ERROR:           desc = "error";           break;
    case Spk_HALT_INDEX_ERROR:     desc = "index error";     break;
    case Spk_HALT_MEMORY_ERROR:    desc = "memory error";    break;
    case Spk_HALT_RUNTIME_ERROR:   desc = "runtime error";   break;
    case Spk_HALT_TYPE_ERROR:      desc = "type error";      break;
    case Spk_HALT_VALUE_ERROR:     desc = "value error";     break;
    default:
        desc = "unknown";
        break;
    }
    return desc;
}


/****/ void SpkHost_Halt(int code, const char *message) {
    if (theInterpreter)
        SpkInterpreter_PrintCallStack(theInterpreter);
    fprintf(stderr, "halt: %s: %s\n", haltDesc(code), message);
    abort(); /* XXX: The proper thing to do is unwind. */
}

/****/ void SpkHost_HaltWithString(int code, SpkUnknown *message) {
    SpkString *str = Spk_CAST(String, message);
    SpkHost_Halt(code, SpkString_AsCString(str));
}

/****/ void SpkHost_VHaltWithFormat(int code, const char *format, va_list args) {
    fprintf(stderr, "halt: %s: ", haltDesc(code));
    vfprintf(stderr, format, args);
    fprintf(stderr, "\n");
    abort(); /* XXX */
}


/*------------------------------------------------------------------------*/
/* base types */

/*****/ int SpkHost_IsInteger(SpkUnknown *op) {
    return Spk_CAST(Integer, op) != (SpkInteger *)0;
}

/*****/ int SpkHost_IsString(SpkUnknown *op) {
    return Spk_CAST(String, op) != (SpkString *)0;
}

/****/ long SpkHost_IntegerAsCLong(SpkUnknown *op) {
    return SpkInteger_AsCLong(Spk_CAST(Integer, op));
}

SpkUnknown *SpkHost_IntegerFromCLong(long value) {
    return (SpkUnknown *)SpkInteger_FromCLong(value);
}

SpkUnknown *SpkHost_FloatFromCDouble(double value) {
    return (SpkUnknown *)SpkFloat_FromCDouble(value);
}

/***/ char *SpkHost_StringAsCString(SpkUnknown *op) {
    SpkString *str = Spk_CAST(String, op);
    return SpkString_AsCString(str);
}

SpkUnknown *SpkHost_StringFromCString(const char *str) {
    return (SpkUnknown *)SpkString_FromCString(str);
}

SpkUnknown *SpkHost_StringFromCStringAndLength(const char *str, size_t len) {
    return (SpkUnknown *)SpkString_FromCStringAndLength(str, len);
}

SpkUnknown *SpkHost_StringConcat(SpkUnknown **a, SpkUnknown *b) {
    return (SpkUnknown *)SpkString_Concat((SpkString **)a, Spk_CAST(String, b));
}


/*------------------------------------------------------------------------*/
/* symbols */

/*****/ int SpkHost_IsSymbol(SpkUnknown *op) {
    return Spk_CAST(Symbol, op) != (SpkSymbol *)0;
}

const char *SpkHost_SymbolAsCString(SpkUnknown *unk) {
    SpkSymbol *sym = Spk_CAST(Symbol, unk);
    if (sym) {
        return SpkSymbol_AsCString(sym);
    }
    return 0;
}

SpkUnknown *SpkHost_SymbolFromCString(const char *str) {
    return (SpkUnknown *)SpkSymbol_FromCString(str);
}

SpkUnknown *SpkHost_SymbolFromCStringAndLength(const char *str, size_t len) {
    return (SpkUnknown *)SpkSymbol_FromCStringAndLength(str, len);
}


/*------------------------------------------------------------------------*/
/* selectors */

const char *SpkHost_SelectorAsCString(SpkUnknown *selector) {
    SpkSymbol *sym = Spk_CAST(Symbol, selector);
    if (sym) {
        return SpkSymbol_AsCString(sym);
    }
    return 0;
}

SpkUnknown *SpkHost_NewKeywordSelectorBuilder(void) {
    return (SpkUnknown *)SpkArray_New(2);
}

/****/ void SpkHost_AppendKeyword(SpkUnknown **builder, SpkUnknown *kw) {
    SpkArray *kwArray;
    size_t size, i;
    SpkUnknown *vd;
    
    kwArray = Spk_CAST(Array, *builder);
    size = SpkArray_Size(kwArray);
    for (i = 0; i < size; ++i) {
        SpkUnknown *item = SpkArray_GetItem(kwArray, i);
        if (item == Spk_uninit) {
            Spk_DECREF(item);
            break;
        }
        Spk_DECREF(item);
    }
    if (i == size) {
        SpkArray *newArray;
        size_t newSize;
        
        newSize = size * 2;
        newArray = SpkArray_New(newSize);
        for (i = 0; i < size; ++i) {
            SpkUnknown *item = SpkArray_GetItem(kwArray, i);
            vd = SpkArray_SetItem(newArray, i, item);
            Spk_DECREF(vd);
            Spk_DECREF(item);
        }
        Spk_DECREF(kwArray);
        kwArray = newArray;
        *builder = (SpkUnknown *)kwArray;
    }
    vd = SpkArray_SetItem(kwArray, i, kw);
    Spk_DECREF(vd);
}

static int compareSymbols(SpkUnknown *const *p1, SpkUnknown *const *p2) {
    SpkSymbol *sym1, *sym2;
    const char *s1, *s2;
    
    sym1 = Spk_CAST(Symbol, *p1);
    sym2 = Spk_CAST(Symbol, *p2);
    if (!sym1 && !sym2)
        return 0;
    if (!sym1)
        return 1;
    if (!sym2)
        return -1;
    s1 = SpkSymbol_AsCString(sym1);
    s2 = SpkSymbol_AsCString(sym2);
    return strcmp(s1, s2);
}

SpkUnknown *SpkHost_GetKeywordSelector(SpkUnknown *builder, SpkUnknown *kw) {
    SpkArray *kwArray;
    SpkSymbol *sym, *selector;
    size_t size, len, i, tally;
    char *str;
    
    kwArray = Spk_CAST(Array, builder);
    size = SpkArray_Size(kwArray);
    len = 0;
    
    if (kw) {
        sym = Spk_CAST(Symbol, kw);
        len += strlen(SpkSymbol_AsCString(sym));
        if (0) {
            /* XXX: The formal arguments must be permuted in the same
               way. */
            SpkArray_Sort(kwArray, &compareSymbols);
        }
    }
    for (i = 0; i < size; ++i) {
        SpkUnknown *item = SpkArray_GetItem(kwArray, i);
        if (item == Spk_uninit) {
            Spk_DECREF(item);
            break;
        }
        sym = Spk_CAST(Symbol, item);
        len += strlen(SpkSymbol_AsCString(sym)) + 1;
        Spk_DECREF(item);
    }
    
    tally = i;
    if (tally > 0)
        ++len;
    
    str = (char *)malloc(len + 1);
    str[0] = '\0';
    
    if (kw) {
        sym = Spk_CAST(Symbol, kw);
        strcat(str, SpkSymbol_AsCString(sym));
        if (tally > 0)
            strcat(str, " ");
    }
    for (i = 0; i < size; ++i) {
        SpkUnknown *item = SpkArray_GetItem(kwArray, i);
        if (item == Spk_uninit) {
            Spk_DECREF(item);
            break;
        }
        sym = Spk_CAST(Symbol, item);
        strcat(str, SpkSymbol_AsCString(sym));
        strcat(str, ":");
        Spk_DECREF(item);
    }
    str[len] = '\0';
    
    selector = SpkSymbol_FromCString(str);
    free(str);
    return (SpkUnknown *)selector;
}


/*------------------------------------------------------------------------*/
/* symbol dictionaries */

SpkUnknown *SpkHost_NewSymbolDict(void) {
    return (SpkUnknown *)SpkIdentityDictionary_New();
}

/****/ void SpkHost_DefineSymbol(SpkUnknown *symbolDict, SpkUnknown *selector, SpkUnknown *method) {
    SpkIdentityDictionary *dict = Spk_CAST(IdentityDictionary, symbolDict);
    SpkIdentityDictionary_SetItem(dict, selector, method);
}

SpkUnknown *SpkHost_SymbolValue(SpkUnknown *symbolDict, SpkUnknown *key) {
    SpkIdentityDictionary *dict = Spk_CAST(IdentityDictionary, symbolDict);
    return SpkIdentityDictionary_GetItem(dict, key);
}

SpkUnknown *SpkHost_FindSymbol(SpkUnknown *symbolDict, SpkUnknown *method) {
    SpkIdentityDictionary *dict = Spk_CAST(IdentityDictionary, symbolDict);
    return SpkIdentityDictionary_KeyWithValue(dict, method);
}

/*****/ int SpkHost_NextSymbol(SpkUnknown *symbolDict,
                               size_t *pPos,
                               SpkUnknown **pKey,
                               SpkUnknown **pValue)
{
    SpkIdentityDictionary *dict;
    
    dict = Spk_CAST(IdentityDictionary, symbolDict);
    return SpkIdentityDictionary_Next(dict, pPos, pKey, pValue);
}


/*------------------------------------------------------------------------*/
/* literal dictionaries */

SpkUnknown *SpkHost_NewLiteralDict(void) {
    /* XXX: IdentityDictionary doesn't really do the job.  It works
     * for Symbols, but it fails to uniquify strings, integers,
     * floats...
     */
    return (SpkUnknown *)SpkIdentityDictionary_New();
}

unsigned int SpkHost_InsertLiteral(SpkUnknown *literalDict, SpkUnknown *literal) {
    SpkIdentityDictionary *dict;
    SpkUnknown *value;
    unsigned int index;
    
    dict = Spk_CAST(IdentityDictionary, literalDict);
    value = SpkIdentityDictionary_GetItem(dict, literal);
    if (value) {
        index = SpkInteger_AsCLong((SpkInteger *)Spk_CAST(Integer, value));
    } else {
        index = SpkIdentityDictionary_Size(dict);
        value = (SpkUnknown *)SpkInteger_FromCLong(index);
        SpkIdentityDictionary_SetItem(dict, literal, value);
    }
    Spk_DECREF(value);
    return index;
}


/*------------------------------------------------------------------------*/
/* arguments  */

/*****/ int SpkHost_IsArgs(SpkUnknown *op) {
    return Spk_CAST(Array, op) != (SpkArray *)0;
}

/**/ size_t SpkHost_ArgsSize(SpkUnknown *args) {
    return SpkArray_Size(Spk_CAST(Array, args));
}

SpkUnknown *SpkHost_GetArg(SpkUnknown *args, size_t index) {
    return SpkArray_GetItem(Spk_CAST(Array, args), index);
}

SpkUnknown *SpkHost_EmptyArgs(void) {
    return (SpkUnknown *)SpkArray_New(0);
}

SpkUnknown *SpkHost_ArgsFromVAList(va_list ap) {
    return (SpkUnknown *)SpkArray_FromVAList(ap);
}

SpkUnknown *SpkHost_GetArgs(SpkUnknown **stackPointer,
                            size_t argumentCount,
                            SpkUnknown *varArgs,
                            size_t skip)
{
    SpkArray *varArgArray;
    
    varArgArray = varArgs ? Spk_CAST(Array, varArgs) : 0;
    return (SpkUnknown *)SpkArray_WithArguments(stackPointer,
                                                argumentCount,
                                                varArgArray,
                                                skip);
}


/*------------------------------------------------------------------------*/
/* i/o */

/*****/ int SpkHost_IsFileStream(SpkUnknown *op) {
    return Spk_CAST(FileStream, op) != (SpkFileStream *)0;
}

/***/ FILE *SpkHost_FileStreamAsCFileStream(SpkUnknown *op) {
    SpkFileStream *stream = Spk_CAST(FileStream, op);
    return SpkFileStream_AsCFileStream(stream);
}


/*------------------------------------------------------------------------*/
/* as yet unclassified */

SpkUnknown *SpkHost_ObjectAsString(SpkUnknown *obj) {
    return Spk_SendMessage(
        theInterpreter,
        obj,
        Spk_METHOD_NAMESPACE_RVALUE,
        Spk_printString,
        Spk_emptyArgs
        );
}

/****/ void SpkHost_PrintObject(SpkUnknown *obj, FILE *stream) {
    SpkUnknown *printObj;
    SpkString *printString;
    char *str;
    
    printObj = SpkHost_ObjectAsString(obj);
    if (!printObj)
        return;
    printString = Spk_CAST(String, printObj);
    if (printString) {
        str = SpkString_AsCString(printString);
        fputs(str, stream);
    }
    Spk_DECREF(printObj);
}
