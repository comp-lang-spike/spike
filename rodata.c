
#include "rodata.h"

#include "host.h"
#include "native.h"
#include "obj.h"
#include <stddef.h>
#include <string.h>


typedef struct { SpkUnknown **p; char *s; }           SymbolTableEntry;
typedef SymbolTableEntry                              SelectorTableEntry;
typedef struct { SpkUnknown **p; long i; }            IntegerTableEntry;
typedef struct { SpkUnknown **p; char *s; size_t n; } StringTableEntry;


/* symbols */
SpkUnknown *Spk_self;


/* selectors */
SpkUnknown *Spk___add__;
SpkUnknown *Spk___addr__;
SpkUnknown *Spk___apply__;
SpkUnknown *Spk___band__;
SpkUnknown *Spk___bneg__;
SpkUnknown *Spk___bor__;
SpkUnknown *Spk___bxor__;
SpkUnknown *Spk___div__;
SpkUnknown *Spk___eq__;
SpkUnknown *Spk___ge__;
SpkUnknown *Spk___gt__;
SpkUnknown *Spk___ind__;
SpkUnknown *Spk___index__;
SpkUnknown *Spk___le__;
SpkUnknown *Spk___lneg__;
SpkUnknown *Spk___lshift__;
SpkUnknown *Spk___lt__;
SpkUnknown *Spk___mod__;
SpkUnknown *Spk___mul__;
SpkUnknown *Spk___ne__;
SpkUnknown *Spk___neg__;
SpkUnknown *Spk___pos__;
SpkUnknown *Spk___pred__;
SpkUnknown *Spk___rshift__;
SpkUnknown *Spk___str__;
SpkUnknown *Spk___sub__;
SpkUnknown *Spk___succ__;
SpkUnknown *Spk__import;
SpkUnknown *Spk__init;
SpkUnknown *Spk__predef;
SpkUnknown *Spk_badExpr;
SpkUnknown *Spk_blockCopy;
SpkUnknown *Spk_cannotReenterBlock;
SpkUnknown *Spk_cannotReturn;
SpkUnknown *Spk_compoundExpression;
SpkUnknown *Spk_doesNotUnderstand;
SpkUnknown *Spk_failOnError;
SpkUnknown *Spk_importModule;
SpkUnknown *Spk_importPackage;
SpkUnknown *Spk_init;
SpkUnknown *Spk_main;
SpkUnknown *Spk_mustBeBoolean;
SpkUnknown *Spk_mustBeSymbol;
SpkUnknown *Spk_mustBeTuple;
SpkUnknown *Spk_new;
SpkUnknown *Spk_noRunnableFiber;
SpkUnknown *Spk_numArgs;
SpkUnknown *Spk_printString;
SpkUnknown *Spk_recursiveDoesNotUnderstand;
SpkUnknown *Spk_redefinedSymbol;
SpkUnknown *Spk_source;
SpkUnknown *Spk_undefinedSymbol;
SpkUnknown *Spk_unknownOpcode;
SpkUnknown *Spk_wrongNumberOfArguments;

SpkUnknown *Spk_Boolean;
SpkUnknown *Spk_False;
SpkUnknown *Spk_True;

SpkSpecialSelector Spk_operSelectors[Spk_NUM_OPER] = {
    { &Spk___succ__,   0 },
    { &Spk___pred__,   0 },
    { &Spk___addr__,   0 }, /* "&x" -- not implemented */
    { &Spk___ind__,    0 }, /* "*x" -- not implemented */
    { &Spk___pos__,    0 },
    { &Spk___neg__,    0 },
    { &Spk___bneg__,   0 },
    { &Spk___lneg__,   0 },
    { &Spk___mul__,    1 },
    { &Spk___div__,    1 },
    { &Spk___mod__,    1 },
    { &Spk___add__,    1 },
    { &Spk___sub__,    1 },
    { &Spk___lshift__, 1 },
    { &Spk___rshift__, 1 },
    { &Spk___lt__,     1 },
    { &Spk___gt__,     1 },
    { &Spk___le__,     1 },
    { &Spk___ge__,     1 },
    { &Spk___eq__,     1 },
    { &Spk___ne__,     1 },
    { &Spk___band__,   1 },
    { &Spk___bxor__,   1 },
    { &Spk___bor__,    1 }
};

SpkSpecialSelector Spk_operCallSelectors[Spk_NUM_CALL_OPER] = {
    { &Spk___apply__,  0 },
    { &Spk___index__,  0 }
};


/* integers */
SpkUnknown *Spk_1;


/* strings */
SpkUnknown *Spk_emptyString; static char emptyString[] = "";
SpkUnknown *Spk_unknownSelector; static char unknownSelector[] = "<unknown>";
SpkUnknown *Spk_unknownSourcePathname; static char unknownSourcePathname[] = "<filename>";


/* arguments */
SpkUnknown *Spk_emptyArgs;


static SymbolTableEntry symbolTable[] = {
    { &Spk_self, "self" },
    { 0 }
};

static SymbolTableEntry selectorTable[] = {
    { &Spk___add__,                       "__add__"                     },
    { &Spk___addr__,                      "__addr__"                    },
    { &Spk___apply__,                     "__apply__"                   },
    { &Spk___apply__,                     "__apply__"                   },
    { &Spk___band__,                      "__band__"                    },
    { &Spk___bneg__,                      "__bneg__"                    },
    { &Spk___bor__,                       "__bor__"                     },
    { &Spk___bxor__,                      "__bxor__"                    },
    { &Spk___div__,                       "__div__"                     },
    { &Spk___eq__,                        "__eq__"                      },
    { &Spk___ge__,                        "__ge__"                      },
    { &Spk___gt__,                        "__gt__"                      },
    { &Spk___ind__,                       "__ind__"                     },
    { &Spk___index__,                     "__index__"                   },
    { &Spk___le__,                        "__le__"                      },
    { &Spk___lneg__,                      "__lneg__"                    },
    { &Spk___lshift__,                    "__lshift__"                  },
    { &Spk___lt__,                        "__lt__"                      },
    { &Spk___mod__,                       "__mod__"                     },
    { &Spk___mul__,                       "__mul__"                     },
    { &Spk___ne__,                        "__ne__"                      },
    { &Spk___neg__,                       "__neg__"                     },
    { &Spk___pos__,                       "__pos__"                     },
    { &Spk___pred__,                      "__pred__"                    },
    { &Spk___rshift__,                    "__rshift__"                  },
    { &Spk___str__,                       "__str__"                     },
    { &Spk___sub__,                       "__sub__"                     },
    { &Spk___succ__,                      "__succ__"                    },
    { &Spk__import,                       "_import"                     },
    { &Spk__init,                         "_init"                       },
    { &Spk__predef,                       "_predef"                     },
    { &Spk_badExpr,                       "badExpr:"                    },
    { &Spk_blockCopy,                     "blockCopy"                   },
    { &Spk_cannotReenterBlock,            "cannotReenterBlock"          },
    { &Spk_cannotReturn,                  "cannotReturn"                },
    { &Spk_compoundExpression,            "compoundExpression"          },
    { &Spk_doesNotUnderstand,             "doesNotUnderstand:"          },
    { &Spk_failOnError,                   "failOnError"                 },
    { &Spk_importModule,                  "importModule"                },
    { &Spk_importPackage,                 "importPackage"               },
    { &Spk_init,                          "init"                        },
    { &Spk_main,                          "main"                        },
    { &Spk_mustBeBoolean,                 "mustBeBoolean"               },
    { &Spk_mustBeSymbol,                  "mustBeSymbol"                },
    { &Spk_mustBeTuple,                   "mustBeTuple"                 },
    { &Spk_new,                           "new"                         },
    { &Spk_noRunnableFiber,               "noRunnableFiber"             },
    { &Spk_numArgs,                       "numArgs"                     },
    { &Spk_printString,                   "printString"                 },
    { &Spk_recursiveDoesNotUnderstand,    "recursiveDoesNotUnderstand:" },
    { &Spk_redefinedSymbol,               "redefinedSymbol:"            },
    { &Spk_source,                        "source"                      },
    { &Spk_undefinedSymbol,               "undefinedSymbol:"            },
    { &Spk_unknownOpcode,                 "unknownOpcode"               },
    { &Spk_wrongNumberOfArguments,        "wrongNumberOfArguments"      },
    
    { &Spk_Boolean,                       "Boolean"                     },
    { &Spk_False,                         "False"                       },
    { &Spk_True,                          "True"                        },
    
    { 0 }
};

static IntegerTableEntry integerTable[] = {
    { &Spk_1, 1 },
    { 0 }
};

static StringTableEntry stringTable[] = {
    { &Spk_emptyString, emptyString, sizeof(emptyString) },
    { &Spk_unknownSelector, unknownSelector, sizeof(unknownSelector) },
    { &Spk_unknownSourcePathname, unknownSourcePathname, sizeof(unknownSourcePathname) },
    { 0 }
};


static int initSymbols(SymbolTableEntry *t) {
    for ( ; t->p; ++t) {
        *t->p = SpkHost_SymbolFromCString(t->s);
        if (!*t->p) {
            return -1;
        }
    }
    return 0;
}

static int initSelectors(SelectorTableEntry *t) {
    for ( ; t->p; ++t) {
        *t->p = Spk_ParseSelector(t->s);
        if (!*t->p) {
            return -1;
        }
    }
    return 0;
}

static int initIntegers(IntegerTableEntry *t) {
    for ( ; t->p; ++t) {
        *t->p = SpkHost_IntegerFromCLong(t->i);
        if (!*t->p) {
            return -1;
        }
    }
    return 0;
}

static int initStrings(StringTableEntry *t) {
    for ( ; t->p; ++t) {
        *t->p = SpkHost_StringFromCStringAndLength(t->s, t->n - 1);
        if (!*t->p) {
            return -1;
        }
    }
    return 0;
}


int Spk_InitSymbols(void) {
    if (initSymbols(symbolTable) < 0)       { return -1; }
    if (initSelectors(selectorTable) < 0)   { return -1; }
    return 0;
}

int Spk_InitReadOnlyData(void) {
    if (initIntegers(integerTable) < 0)     { return -1; }
    if (initStrings(stringTable) < 0)       { return -1; }
    
    Spk_emptyArgs = SpkHost_EmptyArgs();
    if (!Spk_emptyArgs) {
        return -1;
    }
    
    return 0;
}

SpkUnknown *Spk_ParseSelector(const char *methodName) {
    SpkUnknown *selector;
    
    if (strchr(methodName, ':')) {
        /* keyword message */
        SpkUnknown *kw, *builder;
        const char *begin, *end;
        
        if (strchr(methodName, ' ') /*XXX*/ ||
            methodName[strlen(methodName) - 1] != ':') {
            Spk_Halt(Spk_HALT_ASSERTION_ERROR, "invalid method name");
            return 0;
        }
        builder = SpkHost_NewKeywordSelectorBuilder();
        for (begin = end = methodName; *end; begin = end) {
            while (*end != ':')
                ++end;
            kw = SpkHost_SymbolFromCStringAndLength(begin, end - begin);
            SpkHost_AppendKeyword(&builder, kw);
            Spk_DECREF(kw);
            ++end; /* skip ':' */
        }
        selector = SpkHost_GetKeywordSelector(builder, 0);
        Spk_DECREF(builder);
        
    } else {
        selector = SpkHost_SymbolFromCString(methodName);
    }
    
    return selector;
}
