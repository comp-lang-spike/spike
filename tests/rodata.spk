#!/usr/bin/env spike

puts(s) { stdout.printf("%s\n", s); }


hash() {
    obj x, y;
    x = 1.0;
    y = 1;
    if (x === y) {
        // Fixed in r167.
        puts("ERROR: ints and floats hash to the same slots");
    } else {
        puts("int hash OK");
    }
    
    obj s1, s2;
    s1 = "Life is a tomato.";
    s2 = "Life is a tomato.";
    if (s1 !== s2) {
        puts("ERROR: duplicate strings in rodata");
    } else {
        puts("str hash OK");
    }
}


main(argv) {
    hash();
    return 0;
}
