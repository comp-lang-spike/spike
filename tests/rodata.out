int hash OK
str hash OK
